### DISCUSSION TITLE
ECONNRESET Error When Trying to Run NPM Install Behind MITMproxy 9.0.1

### DISCUSSION THREAD

**Discussion by emmettmcdow** [mitmproxy/mitmproxy#6001] [Q&A] [answered]
So I am trying to run `npm install` behind a proxy. Specifically MITMproxy 9.0.1, hosted locally on the same machine as the build machine. I am running the following:
```bash
# Clear cache and artifacts
rm -rf package-lock.json ~/.npm node_modules
# Run Proxy
mitmdump --set stream_large_bodies=1m --ssl-insecure --listen-port 8989 &> stdout.txt &
# build command
npm install --registry=[registry] --force --legacy-peer-deps
```

Which results in the error:
```log
npm ERR! code ECONNRESET
npm ERR! syscall read
npm ERR! errno ECONNRESET
npm ERR! network Invalid response body while trying to fetch [registry]/@angular-devkit%2fbuild-angular: read ECONNRESET
npm ERR! network This is a problem related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2023-03-16T16_07_07_514Z-debug-0.log
```
My `.npmrc` looks like this(running within a docker container for testing purposes, that's why it's using root):
```rc
cafile=/root/.mitmproxy/mitmproxy-ca-cert.pem
https-proxy=https://127.0.0.1:8989
proxy=http://127.0.0.1:8989
```
And the `package.json` looks like this:
```json
{
  "name": "mitmdump-test",
  "version": "0.0.0",
  "private": true,
  "dependencies": {
    "@angular-devkit/build-angular": "^15.1.1",
    "@angular/animations": "^15.1.0",
    "@angular/cli": "^15.1.1",
    "@angular/common": "^15.1.0",
    "@angular/compiler": "^15.1.0",
    "@angular/core": "^15.1.0",
    "@angular/forms": "^15.1.0",
    "@angular/platform-browser": "^15.1.0",
    "@angular/platform-browser-dynamic": "^15.1.0",
    "@angular/router": "^15.1.0"
  }
}
```

#  Environment
## OS
This was tested on Centos8, and Debian Bullseye. This behavior is seen both in containers and traditional VMs.
## Configuration
I discovered that NPM + MITMproxy could cause MITMproxy to barf, as it had too many open file descriptors(in the form of sockets). I got around this by increasing the max file descriptors using `ulimit -n 4096`. This seemed to avoid this problem. 
# Solutions(hacks)
I have identified three actions which fix this. None are acceptable for me to actually do, for one reason or another. And secondly they do not seem to fix the core problem.
1. Downgrading MITMProxy
2. Using the `proxy_debug` option
3. Invoking `npm install` with sudo
4. Accessing the registry using http instead of https

## Downgrading MITMProxy
This Python3/MITMproxy combo works as expected, no `ECONNRESET` error:
```
$ mitmdump --version
Mitmproxy: 5.3.0
Python:    3.6.8
```
But this Python3/MITMproxy combo fails as described above:
```
mitmdump --version
Mitmproxy: 9.0.1
Python:    3.10.4+
```
## Debug logging
If I invoke mitmdump using the `proxy_debug=true`, as so `mitmdump --set stream_large_bodies=1m --set proxy_debug=true --ssl-insecure --listen-port 8989 &> stdout.txt &`, I can get it to work properly. Albeit this option significantly slows things down. I imagine this has something to do with throttling mitmdump?
## Sudo
If you run the `npm install` command with sudo, with the version of Python/mitmdump that typically breaks it, it works properly. I mentioned above that I was testing using a docker container, but in this case, this is only a fix when running outside of root. I've tested both in and out of docker containers. The example above was simply my test that was w/in a docker container. 

## Insecure Registry
Normally I would use https instead of http, but if you switch the registry to use http, it works as intended.
> Note: You might be led to believe this is a ssl-cert permissions issue. It is not. I made sure that the ca file accessed has the correct permissions
# Breaking Version
It looks like this issue was introduced from MITMproxy 6.0.2 -> 7.0.0

**Comment by mhils** [comment]
Without fully looking into this, here are some things you can try:

1. First, check if disabling HTTP/2 fixes things.
2. Second, I would normally recommend `-vvv --set proxy_debug`. But per your description that makes the bug go away, so we are dealing with some sort of nasty race condition.
3. Try setting `https-proxy` to `http://127.0.0.1:8989` (http instead of https)?
4. Finally, maybe setting SSLKEYLOGFILE + inspect traffic with WireShark might be useful to figure out what's going on.

**Comment by emmettmcdow** [comment, marked as answer]
Update: So I went through and tried debugging using the steps outlined by @mhils above. 1-3 were unsuccessful, and nothing about the logs from Wireshark showed anything particularly fishy. But it did cause me to reconsider from where the problem originates. I found [this thread](https://github.com/orgs/nodejs/discussions/44103) about users having similar issues with Node 16 and proxies. Specifically [this comment](https://github.com/orgs/nodejs/discussions/44103#discussioncomment-4727484) that mentions that sometimes npm will incorrectly throw ECONNRESET if certain packets are received out-of-order. I went ahead and tested my case outlined above on npm 14, 16, 18, and 20. v16, the version I had been using, was the only version that broke my case. For the time being, I don't necessarily think this is a problem with MITMproxy.

**Reply by mhils** [reply]
Thanks for digging into this! 🍰

**Comment by Serlena** [comment]
> Update: So I went through and tried debugging using the steps outlined by @mhils above. 1-3 were unsuccessful, and nothing about the logs from Wireshark showed anything particularly fishy. But it did cause me to reconsider from where the problem originates. I found [this thread](https://github.com/orgs/nodejs/discussions/44103) about users having similar issues with Node 16 and proxies. Specifically [this comment](https://github.com/orgs/nodejs/discussions/44103#discussioncomment-4727484) that mentions that sometimes npm will incorrectly throw ECONNRESET if certain packets are received out-of-order. I went ahead and tested my case outlined above on npm 14, 16, 18, and 20. v16, the version I had been using, was the only version that broke my case. For the time being, I don't necessarily think this is a problem with MITMproxy.

Hi, I am a junior developer facing the same issue. hence how to resolve this by the way, the node version i am using is 16.20.1 and it keeps throwing this error, i have tried `npm install --legacy-peer-deps` still the same error, appreciate your help please.

![image](https://github.com/mitmproxy/mitmproxy/assets/81607181/edc808b2-fc19-4b80-b3be-86ac00b5ac0c)

**Reply by salvadorsystems** [reply]
I also have the same error, did you solve the error?

**Reply by Mhd-ysn7** [reply]
did you solve it?

**Reply by salvadorsystems** [reply]
Yes, this worked for me:
1.do a pull of the project
2. update the angular version with the following command
 ng update @angular/core@12.0.1
3. Update so that any version runs on your PC
 ng install
4. -- Command to run angular on pc and start it
set NODE_OPTIONS=--openssl-legacy-provider
ng serve

**Reply by codingwithsaleem** [reply]
to resolve this issue 

delete the package lock and node module folder 

then add the nodejs engine in package.json file 

"engines": {
    "node": "20.x",
    "npm": "10.x"
  }
  
  then run npm install and push the changes
